pkg/controlplane/controller/legacytokentracking/controller.go

```diff
@@ -0,0 +1,211 @@
+		options.FieldSelector = fields.OneTermEqualSelector("metadata.name", configMapName).String()
+	})
+
+	c := &Controller{
+		configMapClient:   cs.CoreV1(),
+		namespaceClient:   cs.CoreV1(),
+		namespaceExists:   false,
+		queue:             workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "legacy_token_tracking_controller"),
+		configMapInformer: informer,
+		configMapCache:    informer.GetIndexer(),
+		configMapSynced:   informer.HasSynced,
+		enabled:           utilfeature.DefaultFeatureGate.Enabled(kubefeatures.LegacyServiceAccountTokenTracking),
+	}
+
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			configMap := obj.(*corev1.ConfigMap)
+			c.enqueue(configMap)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			configMap := newObj.(*corev1.ConfigMap)
+			c.enqueue(configMap)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.freeze()
+			configMap := obj.(*corev1.ConfigMap)
+			c.enqueue(configMap)
+		},
+	})
+
+	return c
+}
+
+func (c *Controller) enqueue(configMap *corev1.ConfigMap) {
+	key, err := cache.MetaNamespaceKeyFunc(configMap)
+	if err != nil {
+		klog.Errorf("Failed to get key for ConfigMap: %v", configMap)
+		return
+	}
+	c.queue.Add(key)
+}
+
+func (c *Controller) freeze() {
+	// when delete happens and the controller is enabled to create the
+	// configmap, this means there are multiple controller instances and
+	// at least one of them is not enabled to create the configmap. in
+	// such case, we should stop creating configmap in the next 30 mins.
+	// eventually, all controller instances should be either enabled or
+	// not and this condition will not be entered unless further change
+	// happens.
+	if c.enabled {
+		c.freezeUntil = time.Now().Add(freezePeriod)
+	}
+}
+
+// Run starts the controller sync loop.
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Info("Starting legacy_token_tracking_controller")
+	if !cache.WaitForNamedCacheSync("tokens", stopCh, c.configMapSynced) {
+		return
+	}
+
+	go c.configMapInformer.Run(stopCh)
+	go wait.Until(c.runWorker, time.Second, stopCh)
+
+	<-stopCh
+	klog.Info("Ending legacy_token_tracking_controller")
+}
+
+func (c *Controller) runWorker() {
+	for c.processNext() {
+	}
+}
+
+func (c *Controller) processNext() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	if err := c.syncConfigMap(key.(string)); err != nil {
+		utilruntime.HandleError(fmt.Errorf("while syncing ConfigMap %q, err: %w", key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	}
+
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *Controller) syncConfigMap(key string) error {
+	_, exists, err := c.configMapCache.GetByKey(key)
+	if err != nil {
+		return err
+	}
+
```

Comment:
this function would be easier to follow if the enabled and disabled branches were completely separate

something like:

```go
switch {
case c.enabled:
  if !exists {
    ... attempt create ...
  } else if exists but timestamp data is not valid {
    ... attempt update with valid timestamp data ...
  }

case !c.enabled:
  if !exists {
    return nil
  }
  ... attempt delete ...
}
```

Reply:
done